aws/lambda/pytorch-auto-revert/pytorch_auto_revert/signal.py

```diff
@@ -0,0 +1,291 @@
+            None is "Maybe", meaning the result depends on the resolution of the existing pending job.
+        """
+        if len(self.commits) < 2:
+            return False
+
+        maybe = False
+
+        # Iterate through commits, looking for a newer commit with failure
+        for i in range(0, len(self.commits) - 1):
+            nc = self.commits[i]
+
+            # Check all older commits before this one
+            for j in range(i + 1, len(self.commits)):
+                oc = self.commits[j]
+                # Check if this older commit has two successful jobs
+                oc_successes = oc.events_by_status(SignalStatus.SUCCESS)
+                oc_pending = oc.events_by_status(SignalStatus.PENDING)
+
+                if len(oc_successes) >= 2:
+                    # Check if the failure of the newer commit is between the two successes of the older commit
+                    # Events are ordered by time within each commit, oldest events first
+                    for e in nc.events:
+                        if (
+                            oc_successes[0].started_at
+                            < e.started_at
+                            < oc_successes[-1].started_at
+                        ):
+                            if e.is_failure:
+                                # We have a sandwich pattern
+                                return True
+                            elif e.is_pending:
+                                # We have a pending job, possible pattern, cannot confirm yet
+                                maybe = True
+
+                elif (
+                    len(oc_successes) == 1
+                    and len(oc_pending) > 1
+                    and oc_successes[0].started_at < oc_pending[-1].started_at
+                ):
+                    # If there is only one success and multiple pending jobs, we cannot confirm the sandwich
+                    if any(
+                        oc_successes[0].started_at
+                        < e.started_at
+                        < oc_pending[-1].started_at
+                        and (e.is_failure or e.is_pending)
+                        for e in nc.events
+                    ):
+                        maybe = True
+
+        return None if maybe else False
+
+    def has_loose_autorevert_pattern(self) -> bool:
+        """
+        Checks if there is subsequence of commits, where:
+        - there are two commits with a failure (not necessarily consecutive)
+        - there is at least one older commit with a success (not necessarily immediate predecessor)
+        :return:
+            True if such a pattern exists, False otherwise.
+        """
+        if len(self.commits) < 3:
+            return False
+
+        # Check for two newer commits with failure and one older commit with success
+        found_failures = 0
+        for c in self.commits:
```

Comment:
things that I suspect are wrong:

1 - we don't care about any failure, we care if all events are failing (in this case any failure would do)
2 - We ignore PENDING. If there is a pending job we should wait it resolve.
3 - it only detects the occurrence of a pattern in the self.commits[1] as suspected commit. Not sure if this was the intention.

Reply:
per our discussion: the current logic is intentional

for simplicity and composability with other invariant detectors we're ignoring the case when flakiness recovers within the short lookback window

the current logic is conservative: if we flag signal as flaky, we not using it.

Reply:
per our discussion, simplified this method to only check for the presence of any success (now called `has_successes`)